031-custom_authentication.md

```diff
@@ -0,0 +1,97 @@
+# Custom Authentication
+
+This proposal focuses on supporting custom authentication in the Strimzi operator.
+
+## Current situation
+
+Currently, numerous authentication methods are [supported](https://github.com/strimzi/strimzi-kafka-operator/tree/0.25.0/api/src/main/java/io/strimzi/api/kafka/model/authentication) in Strimzi, and allows the user to configure which one to use. In addition, in the case a custom *authorizer* needs to be used, this is also currently [supported](https://github.com/strimzi/strimzi-kafka-operator/blob/0.25.0/cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaBrokerConfigurationBuilder.java#L541-L549). 
+
+However, there is no ability to specify custom authentication, which is what this proposal will focus on.
+
+## Motivation
+
+Supporting custom authentication allows greater flexibility for users of Strimzi who may have bespoke/proprietary authentication requirements. This would greatly broaden user adoption of the operator, as it would also support more common authentication schemes, such as [Amazon MSK](https://docs.aws.amazon.com/msk/latest/developerguide/security_iam_service-with-iam.html), and attract users who have much more stringent requirements (i.e. proprietary auth). 
+
+The “cost” of supporting such feature should be minimal as well, as the expectation for custom authentication passes the complexity onto the user. Strimzi should strictly support the passing of configuration values, and nothing else. It’s up to the user to ensure their authn/z schemes can be integrated with Strimzi. Considering this also requires custom built Strimzi Kafka images in order to load custom classes on to the classpath, it’s also expected that will only really apply to highly technical/pro users.
```

Comment:
There are cons to this:

* If we add support for this in the CRD schema then users can configure clusters with a customer authenticator but the operator has no way of validating that the broker image actually has those classes. Until now the operator and apiserver schema validation together have done a reasonable job of validating CR instances. 

* It's easy for more technical users to make use of this, but if they keep that to themselves then the community doesn't benefit. There would likely end up being multiple "technical users" each with their own MSK authentication classes (so they don't get the benefit of there being a single well-maintained implementation), and less sophisticated users don't get any benefit. In other words the bar remains high even when someone scratches their own itch. Obviously there are limited options for doing better:
    * Building support for each authenticator into the Strimzi images. This doesn't feel great to me. While it lowers the bar for end users it raises a different bar: Getting an authenticator into Strimzi is now an impediment.
    * Add support for building broker images, using the Strimzi images as a base, but adding plugin jar files. Similar to what we're doing with the ability to build `KafkaConnect` images which include connector plugins. This would open the door to being able to configure other plugins, such as policy plugins, custom quota managers, logging classes and so on. It would make it easier for less technical users to make use of plugins, and allows communities to grow around those plugins. So there could be a single open source MSK plugin which everyone could easily use. It doesn't preclude users writing their own plugins in private, either, but makes them easier to deploy because they don't have to engage with the Strimzi build system. Obviously this is not a trivial piece of work, but I'd be interested to hear how crazy people thought it was.

Reply:
```If we add support for this in the CRD schema then users can configure clusters with a customer authenticator but the operator has no way of validating that the broker image actually has those classes. Until now the operator and apiserver schema validation together have done a reasonable job of validating CR instances.```

Yeah this is true, and unfortunately, unavoidable. However, this is already the case with the CustomAuthorizer already. 

```It's easy for more technical users to make use of this, but if they keep that to themselves then the community doesn't benefit. There would likely end up being multiple "technical users" each with their own MSK authentication classes (so they don't get the benefit of there being a single well-maintained implementation), and less sophisticated users don't get any benefit. In other words the bar remains high even when someone scratches their own itch. Obviously there are limited options for doing better:```
Agreed that the incentive structure needs to be put-in place to ensure users are contributing back to the community, especially the more technical ones. However, one thing to be mindful is not gatekeeping too hard, as otherwise users will just end up forking, and adding their custom auth anyways. I do think a soft-measure that can be taken here, is that rather the explicitly supporting custom auth in the documentation, simply state Amazon MSK support (+ others which we've images for). Hopefully this lead to the workflow of the user filling an issue asking about auth scheme X, which then a boilerplate response can be given on how-to add support. 

```
Add support for building broker images, using the Strimzi images as a base, but adding plugin jar files. Similar to what we're doing with the ability to build KafkaConnect images which include connector plugins. This would open the door to being able to configure other plugins, such as policy plugins, custom quota managers, logging classes and so on. It would make it easier for less technical users to make use of plugins, and allows communities to grow around those plugins. So there could be a single open source MSK plugin which everyone could easily use. It doesn't preclude users writing their own plugins in private, either, but makes them easier to deploy because they don't have to engage with the Strimzi build system. Obviously this is not a trivial piece of work, but I'd be interested to hear how crazy people thought it was.
```
You could take this a step further, and have the jar just be dynamically downloaded at runtime/pod launch. Although this runs counter to exactly why you've images in the first place, and would make rolls much longer.